conctr.agent.lib.nut

```diff
@@ -195,18 +232,296 @@ class Conctr {
+                // Sweet, we want to extract it out, itll be
+                // chilling just before the /r/n lets find it
+                local eos = chunks.find(STREAM_TERMINATOR);
+                // Got it! 
+                if (eos != null) {
+                    // Pull it out
+                    contentLength = chunks.slice(0, eos);
+                    try {
+                        contentLength = contentLength.tointeger();
+                    } catch (e) {
+                        server.error(e);
+                    }
+
+                    // Leave the rest of the msg
+                    chunks = chunks.slice(eos + STREAM_TERMINATOR.len());
+                }
+            }
+
+            if (contentLength != null && chunks.len() >= contentLength) {
+                // Got a full msg, process it!
+                _processData(chunks.slice(0, contentLength), cb);
+
+                // Get any partial chunks if any and keep waiting for the end of new message
+                chunks = chunks.slice(contentLength + STREAM_TERMINATOR.len());
+                contentLength = null;
+            }
+        }
+
+        headers["Content-Type"] <- "application/json";
+        headers["Connection"] <- "keep-alive";
+        headers["Transfer-encoding"] <- "chunked";
+        headers["Authorization"] <- _headers["Authorization"]
+        payload["topics"] <- topics;
+
+        // Check there isnt a current connection, close it if there is.
+        if (_pollingReq) _pollingReq.cancel();
+
+        _pollingReq = http.post(_pubSubEndpoints[action], headers, http.jsonencode(payload));
+
+        // Call callback directly when not chucked response, handle chuncking in second arg to sendAsync
+        _pollingReq.sendasync(_doneCb.bindenv(this), _streamCb.bindenv(this));
+    }
+
+
+    // 
+    // Unsubscribe to a single/list of topics
+    // 
+    function unsubscribe() {
+        if (_pollingReq) _pollingReq.cancel();
+        _pollingReq = null;
+    }
+
+
+
+    // 
+    // Publishes a message to the correct url.
+    // @param  {String}   relativeUrl   relative url that the message should be posted to
+    // @param  {Table}    receivers     Object with either device_ids or topics
+    // @param  {[type]}   msg           Data to be sent to be published
+    // @param  {[type]}   contentType   Header specifying the content type of the msg
+    // @param  {Function} cb            Function called on completion of publish request
+    // 
+    function _publish(relativeUrl, receivers, msg, contentType = null, cb = null) {
+        local action = "publish";
+        local headers = {};
+        local reqTime = time();
+        local payload = { "msg": msg };
+
+        if (typeof contentType == "function") {
+            cb = contentType;
+            contentType = null;
+        }
+
+        if ("topics" in receivers) {
+            payload["topics"] <- receivers.topics;
+        } else if ("device_ids" in receivers) {
```

Comment:
What if both `topics` and `device_ids` are provided?

Reply:
topics takes precident, this function is not called directly, it is called by parent functions that allow either topics or device id not both.